webapp/src/js/bootstrapper/initial-replication.js

```diff
@@ -0,0 +1,190 @@
+  return remoteDocIdsRevs.filter(({ id, rev }) => !localDocs[id] || localDocs[id] !== rev);
+};
+
+const getDownloadList = async (localDb) => {
+  const response = await utils.fetchJSON('/initial-replication/get-ids');
+
+  docIdsRevs = await getMissingDocIdsRevsPairs(localDb, response.doc_ids_revs);
+  lastSeq = response.last_seq;
+  remoteDocCount = response.doc_ids_revs.length;
+
+  if (response.warn) {
+    await displayTooManyDocsWarning(response);
+  }
+};
+
+const getLocalDocList = async (localDb) => {
+  const response = await localDb.allDocs();
+
+  const localDocMap = {};
+  response.rows.forEach(row => localDocMap[row.id] = row.value && row.value.rev);
+  return localDocMap;
+};
+
+const getDocsBatch = async (remoteDb, localDb) => {
+  const batch = docIdsRevs.splice(0, BATCH_SIZE);
+
+  if (!batch.length) {
+    return;
+  }
+
+  const res = await remoteDb.bulkGet({ docs: batch, attachments: true });
+  const docs = res.results
+    .map(result => result.docs && result.docs[0] && result.docs[0].ok)
+    .filter(doc => doc);
+  await localDb.bulkDocs(docs, { new_edits: false });
+};
+
+const downloadDocs = async (remoteDb, localDb) => {
+  setUiStatus('FETCH_INFO', { count: remoteDocCount - docIdsRevs.length, total: remoteDocCount });
+  do {
+    await getDocsBatch(remoteDb, localDb);
+    setUiStatus('FETCH_INFO', { count: remoteDocCount - docIdsRevs.length, total: remoteDocCount });
+  } while (docIdsRevs.length > 0);
+};
+
+const writeCheckpointers = async (remoteDb, localDb) => {
+  const replicateFromPromise = localDb.replicate.from(remoteDb, {
+    live: false,
+    retry: false,
+    heartbeat: 10000,
+    timeout: 1000 * 60 * 10, // try for ten minutes then give up,
+    query_params: { initial_replication: true },
+    since: lastSeq,
+  });
+
+  replicateFromPromise.on('change', () => replicateFromPromise.cancel());
+  await replicateFromPromise;
+
+  const localInfo = await localDb.info();
+  await localDb.replicate.to(remoteDb, {
+    since: localInfo.update_seq,
+  });
+};
+
+const replicate = async (remoteDb, localDb) => {
+  setUiStatus('LOAD_APP');
+
+  await startInitialReplication(localDb);
+
+  setUiStatus('POLL_REPLICATION');
+  await getDownloadList(localDb);
+  await downloadDocs(remoteDb, localDb);
+  await writeCheckpointers(remoteDb, localDb);
+
+  await completeInitialReplication(localDb);
+};
+
+const getReplicationLog = async (localDb) => {
+  try {
+    return await localDb.get(INITIAL_REPLICATION_LOG);
+  } catch (err) {
+    return null;
+  }
+};
+const isReplicationNeeded = async (localDb, userCtx) => {
+  const requiredDocs = [
+    '_design/medic-client',
+    'settings',
+    `org.couchdb.user:${userCtx.name}`,
+  ];
+  const results = await localDb.allDocs({ keys: requiredDocs });
+  const errors = results.rows.filter(row => row.error);
+  const hasRequiredDocs = !errors.length;
+
+  const replicationLog = await getReplicationLog(localDb);
+
+  if (!hasRequiredDocs) {
+    return true;
+  }
+
```

Comment:
Rearrange this for performance reasons...

```suggestion
  if (!hasRequiredDocs) {
    return true;
  }
  const replicationLog = await getReplicationLog(localDb);
```